Refactor poll_oneoff on *nix

kubkon opened the pull request:
This PR introduces a couple of changes/fixes:
* it annotates `log::debug!` messages with "host" to differentiate
  between file descriptors stored on the WASI side (aka the wrappers)
  and those managed by the host (aka the wrapped)
* it fixes CraneStation/wasmtime#440, i.e., incorrect passing of
  file descriptor to `poll_oneoff` where currently errenously we
  pass in the wrapper instead of the wrapped value
* it adds a couple more `log::debug!` macros calls for easier future
  debugging
* it refactors `poll_oneoff` for easier maintenance in the future (hopefully!)

kubkon committed:
Fixes CraneStation/wasmtime#440

This commit introduces a couple of changes/fixes:
* it annotates `log::debug!` messages with "host" to differentiate
  between file descriptors stored on the WASI side (aka the wrappers)
  and those managed by the host (aka the wrapped)
* it fixes CraneStation/wasmtime#440, i.e., incorrect passing of
  file descriptor to `poll_oneoff` where currently errenously we
  pass in the wrapper instead of the wrapped value
* it adds a couple more `log::debug!` macros calls for easier future
  debugging

kubkon committed:
Add partial refactorting to poll_oneoff

This commit lays the groundwork for more clean up to come in
subsequent commits.

kubkon committed:
Finalise refactoring of `poll_oneoff`

kubkon commented:
OK, this PR is ready for review. ~~Note however that it is (unfortunately) expected to fail as it makes breaking API changes which still need to be addressed in `wasmtime`.~~ The PR is expected to pass the tests now :-)

Additionally, this PR is blocked by CraneStation/wasi-misc-tests#38 (i.e., it really should include `misc_testsuite` fast-forward to the latest upstream when CraneStation/wasi-misc-tests#38 lands).

marmistrz commented on src/hostcalls_impl/misc.rs:
```diff
@@ -205,9 +205,15 @@ pub(crate) fn poll_oneoff(
+    let mut events_count = 0;
+    for (i, event) in events.into_iter().enumerate() {
+        output_slice[i] = enc_event(event);
+        events_count += 1;
```
You can just use `events.len()` and avoid mutability. And please avoid the indices, they're not necessary here (the original code also avoids them). See also how it's done in #140.

kubkon replied:
Good point on `events.len()`, thanks! I'm less clear on "avoiding the indices". Are you saying that calling

```rust
*output_slice_cur.next().unwrap() = enc_event(event);
```

is safer than using indices because AFAIK, in this case, both will panic if we exceed the range. Or is it a stylistic concern?

marmistrz replied:
Both. Probably with an `expect` instead of `unwrap`. It's cleaner and it's not possible to use the wrong variable as the index.

marmistrz commented on src/sys/unix/hostcalls_impl/misc.rs:
```diff
@@ -70,78 +69,127 @@ pub(crate) fn clock_time_get(clock_id: host::__wasi_clockid_t) -> Result<host::_
-        })
-        .min_by_key(|event| event.delay);
+    wasi_ctx: &WasiCtx,
+    input: impl Iterator<Item = Result<host::__wasi_subscription_t>>,
```
Can we accept `Vec<FdEventData>` and  `Option<ClockEventData>` for timeout as in #140? The conversion code cross-platform code and should not reside in this module.

kubkon replied:
~~Ah, good idea. Let me refactor it following #140 then!~~ Actually, on second thought, I'd like to keep this PR focused on fixing a bug in `poll_oneoff`. Would you mind handling more refactoring in a subsequent PR?

marmistrz replied:
I'd prefer to have one decent refactor commit, but if this is a major blocker and should be merged ASAP, I'm fine with finishing the refactor in a subsequent PR.

kubkon replied:
I guess we're not in such a hurry, so let's see how we get on with it :-)

kubkon replied:
OK, I've now had a closer look at this. We cannot accept `Vec<FdEventData>` as this will be host specific (on *nix, we deal with file descriptors, on Windows it will be handles). However, if we modify `FdEventData` to contain our `Descriptor` struct, then this should be possible to refactor out.

marmistrz commented on src/sys/unix/hostcalls_impl/misc.rs:
```diff
@@ -70,78 +69,127 @@ pub(crate) fn clock_time_get(clock_id: host::__wasi_clockid_t) -> Result<host::_
-                Some(FdEventData {
-                    fd: unsafe { event.u.fd_readwrite.fd } as c_int,
+    let mut timeout: Option<ClockEventData> = None;
+    let mut fd_events = Vec::new();
```
See the comment above.

marmistrz commented on src/sys/unix/hostcalls_impl/misc.rs:
```diff
@@ -70,78 +69,127 @@ pub(crate) fn clock_time_get(clock_id: host::__wasi_clockid_t) -> Result<host::_
         })
         .collect();
+
     let timeout = timeout.map(|ClockEventData { delay, userdata }| ClockEventData {
```
We can remove this map and just use `try_into` as in #140.

marmistrz commented on src/sys/unix/hostcalls_impl/misc.rs:
```diff
@@ -161,116 +209,106 @@ fn wasi_clock_to_relative_ns_delay(
     } else {
         // shouldn't happen
-        0
+        vec![]
```
Then accept `ClockEventData`, not an `Option<ClockEventData>`. Accepting `Option<ClockEventData>` suggests that `None` is a valid value. See how #140 does it.

marmistrz reviewed (changes requested):
I closed #140, let's keep all the refactoring in one PR.

kubkon commented:
> I closed #140, let's keep all the refactoring in one PR.

That's my bad at poorly synchronising efforts between the two of us. Apologies!

dunnock commented on src/sys/unix/hostcalls_impl/misc.rs:
```diff
@@ -70,78 +69,127 @@ pub(crate) fn clock_time_get(clock_id: host::__wasi_clockid_t) -> Result<host::_
+                let wasi_fd = unsafe { event.u.fd_readwrite.fd };
+                let fd = unsafe {
+                    wasi_ctx
+                        .get_fd_entry(wasi_fd, host::__WASI_RIGHT_FD_READ, 0)?
```
This line seems throwing error when at least one bad wasi file descriptor supplied. Should not it rather return event for relevant record with error code EBADF? Since there might be mix of polls subscriptions and some might be valid, while some invalid it is important to return those which are valid too.

I've added relevant test https://github.com/kubkon/wasi-misc-tests/pull/2/commits/ed16186b8d4a762020083c8b36d91b12a96f9580

kubkon replied:
Yep, I see it now, thanks!

kubkon replied:
Now that I've included the smoke test CraneStation/wasi-misc-tests#38 in this PR, this seems to be the only (I hope I didn't miss anything!) major thing left before this PR is ready for merging.

kubkon replied:
Here's a question about the spec; would love your comment about this one @sunfishcode. From what I see in the implementation and the spec, the user can link subscriptions with events by the field `userdata` which is present in both and set to the same, user-specified value. Does that mean that we don't have to preserve the order of events relative to the order of the (input) subscriptions? I guess this would make sense especially since in the spec we do say "concurrently poll" [WASI-api#__wasi_poll_oneoff]. And a follow-up question. If indeed `userdata` is used as an identifier for `(subscription, event)` pair, we (purposely) do not enforce uniqueness of `userdata` among the input subscriptions which could very well lead to collisions in the output events, but I guess it is up to the client to enforce the uniqueness?

[WASI-api#__wasi_poll_oneoff]: https://github.com/CraneStation/wasmtime/blob/master/docs/WASI-api.md#__wasi_poll_oneoff

kubkon replied:
OK, I think I now have convinced myself that the order in general will not be preserved nor is expected to be.

kubkon committed:
Fix compilation error on Windows

kubkon committed:
Address majority of suggestions and refactor

Co-authored-by: Marcin Mielniczuk <marmistrz.dev@zoho.eu>

kubkon committed:
Add poll_oneoff test case

marmistrz commented:
@kubkon can we have a more descriptive PR title?

Commit:
Leave timeout in nanoseconds in ClockEventData

Instead of converting the timeout value from nanoseconds to
milliseconds in the host-independent impl, move the conversion
to *nix-specific impl as the conversion is currently only warranted
by the POSIX `poll` syscall.

Commit:
Don't fail immediately on bad descriptor

If the user specifies an invalid descriptor inside a subscription,
don't fail immediately but rather generate an event with the thrown
WASI error code, and continue with the remaining, potentially
correct subscriptions.

kubkon commented:
I think we're in a pretty good shape with this now, so I'm gonna go ahead and merge this. If you guys feel that we should still modify something, or something isn't clear, please feel free to submit an issue, and I'll be happy to fix anything that is remaining :-)
